File: core/src/weights.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const Config = @import("config.zig").Config;

/// Struct defines the weights of moondream
/// All weights are transposed
/// Naming convention :
/// "t_" prefix : model (phi 1.5)
/// "v_" prefix : vision_model (vision encoder)
/// "_w" suffix : weights
/// "_b" suffix : biases
const Weights = struct {
    const Self = @This();
    // Slices for specific weights

    ///Text model start///
    word_token_embedding: []f32, // (dim, vocab)

    // Transformer layer start

    // attn layer norm
    t_ln_w: []f32, // (layer, dim)
    t_ln_b: []f32, // (layer, dim)
    // attn qkv
    t_Wqkv_w: []f32, // (layer, dim, n_heads*head_dim*3)
    t_Wqkv_b: []f32, // (layer, n_heads*head_dim*3)
    // output
    t_out_proj_w: []f32, // (layer, seqlen, dim)
    t_out_proj_bias: []f32, // (layer, dim)
    // fully connected
    t_fc1_w: []f32, // (layer, hidden_dim, dim)
    t_fc1_b: []f32, // (layer, hidden_dim)
    t_fc2_w: []f32, // (layer, dim, hidden_dim)
    t_fc2_b: []f32, // (layer, dim)

    //Transformer layer end //

    // lm head
    t_linear_w: []f32, //(vocab, dim)
    t_linear_b: []f32, //(vocab)
    t_ln_out_w: []f32, //(dim)
    t_ln_out_b: []f32, //(dim)
    //Text model end///

    // Vision model start //

    // combining patch embeddngs and pos
    v_patch_embedding_linear_w: []f32, // (vit_dim, patch * patch * channels)
    v_patch_embedding_linear_b: []f32, // (vit_dim)
    v_pos_embedding: []f32, // (1, (img_dim/patch_dim)^2, vit_dim)

    /// Vision Transformer Start
    // attention qkv
    v_Wqkv_w: []f32, // (vit_dim, vit_dim*3)
    v_Wqkv_b: []f32, // (vit_dim * 3)

    //attn out
    v_out_proj_w: []f32, // (vit_dim, vit_dim)
    v_out_proj_b: []f32, // (vit_dim)

    //ViT fc
    v_fc1_w: []f32, // (hidden_features, vit_dim)
    v_fc1_b: []f32, // (hidden_features)
    v_fc2_w: []f32, // (vit_dim, hidden_features)
    v_fc2_b: []f32, // (vit_dim)

    //ViT norm
    v_norm1_w: []f32, // (layer, hidden_features)
    v_norm1_b: []f32, // (layer, hidden_features)
    v_norm2_w: []f32, // (layer, hidden_features)
    v_norm2_b: []f32, // (layer, hidden_features)

    // Vision Transformer End

    //norm
    v_norm_out_w: []f32, // (hidden_features)
    v_norm_out_b: []f32, // (hidden_features)

    // projection
    v_proj_fc1_w: []f32, // (hidden_dim, hidden_features * 2)
    v_proj_fc1_b: []f32, // (hidden_dim)

    v_proj_fc2_w: []f32, // (hidden_features*2, hidden_dim)
    v_proj_fc2_b: []f32, // (hidden_features)

    fn init(config: Config, filename: []const u8, allocator: Allocator) !Weights {

        // Set up slices for each weight

        const sizes = calculateSizes(config);
        const num_weights = calculateTotalSize(sizes);
        const file = try std.fs.cwd().openFile(filename, .{});
        defer file.close();

        const file_size = try file.getEndPos();
        if (file_size != num_weights * @sizeOf(f32)) {
            std.debug.print("Actual file size = {} \n", .{file_size});
            std.debug.print("Estimated file size = {} \n", .{num_weights * @sizeOf(f32)});
            return error.UnexpectedFileSize;
        }
        std.debug.print("{s} read successfully with {} parameters. \n", .{ filename, num_weights });
        var data = try allocator.alloc(f32, num_weights);
        defer allocator.free(data);

        const bytes_read = try file.readAll(std.mem.sliceAsBytes(data));
        if (bytes_read != file_size) {
            return error.IncompleteRead;
        }

        // Memory mapping of slices
        // Set slices for each weight
        var self: Weights = undefined;
        var offset: usize = 0;
        // text model
        self.word_token_embedding = data[offset .. offset + sizes.word_token_embedding];
        offset += sizes.word_token_embedding;

        self.t_ln_w = data[offset .. offset + sizes.t_ln_w];
        offset += sizes.t_ln_w;

        self.t_ln_b = data[offset .. offset + sizes.t_ln_b];
        offset += sizes.t_ln_b;

        self.t_Wqkv_w = data[offset .. offset + sizes.t_Wqkv_w];
        offset += sizes.t_Wqkv_w;

        self.t_Wqkv_b = data[offset .. offset + sizes.t_Wqkv_b];
        offset += sizes.t_Wqkv_b;

        self.t_out_proj_w = data[offset .. offset + sizes.t_out_proj_w];
        offset += sizes.t_out_proj_w;

        self.t_out_proj_bias = data[offset .. offset + sizes.t_out_proj_bias];
        offset += sizes.t_out_proj_bias;

        self.t_fc1_w = data[offset .. offset + sizes.t_fc1_w];
        offset += sizes.t_fc1_w;

        self.t_fc1_b = data[offset .. offset + sizes.t_fc1_b];
        offset += sizes.t_fc1_b;

        self.t_fc2_w = data[offset .. offset + sizes.t_fc2_w];
        offset += sizes.t_fc2_w;

        self.t_fc2_b = data[offset .. offset + sizes.t_fc2_b];
        offset += sizes.t_fc2_b;

        self.t_linear_w = data[offset .. offset + sizes.t_linear_w];
        offset += sizes.t_linear_w;

        self.t_linear_b = data[offset .. offset + sizes.t_linear_b];
        offset += sizes.t_linear_b;

        self.t_ln_out_w = data[offset .. offset + sizes.t_ln_out_w];
        offset += sizes.t_ln_out_w;

        self.t_ln_out_b = data[offset .. offset + sizes.t_ln_out_b];
        offset += sizes.t_ln_out_b;

        // vision model

        self.v_patch_embedding_linear_w = data[offset .. offset + sizes.v_patch_embedding_linear_w];
        offset += sizes.v_patch_embedding_linear_w;

        self.v_patch_embedding_linear_b = data[offset .. offset + sizes.v_patch_embedding_linear_b];
        offset += sizes.v_patch_embedding_linear_b;

        self.v_pos_embedding = data[offset .. offset + sizes.v_pos_embedding];
        offset += sizes.v_pos_embedding;

        self.v_Wqkv_w = data[offset .. offset + sizes.v_Wqkv_w];
        offset += sizes.v_Wqkv_w;

        self.v_Wqkv_b = data[offset .. offset + sizes.v_Wqkv_b];
        offset += sizes.v_Wqkv_b;

        self.v_out_proj_w = data[offset .. offset + sizes.v_out_proj_w];
        offset += sizes.v_out_proj_w;

        self.v_out_proj_b = data[offset .. offset + sizes.v_out_proj_b];
        offset += sizes.v_out_proj_b;

        self.v_fc1_w = data[offset .. offset + sizes.v_fc1_w];
        offset += sizes.v_fc1_w;

        self.v_fc1_b = data[offset .. offset + sizes.v_fc1_b];
        offset += sizes.v_fc1_b;

        self.v_fc2_w = data[offset .. offset + sizes.v_fc2_w];
        offset += sizes.v_fc2_w;

        self.v_fc2_b = data[offset .. offset + sizes.v_fc2_b];
        offset += sizes.v_fc2_b;

        self.v_norm1_w = data[offset .. offset + sizes.v_norm1_w];
        offset += sizes.v_norm1_w;

        self.v_norm1_b = data[offset .. offset + sizes.v_norm1_b];
        offset += sizes.v_norm1_b;

        self.v_norm2_w = data[offset .. offset + sizes.v_norm2_w];
        offset += sizes.v_norm2_w;

        self.v_norm2_b = data[offset .. offset + sizes.v_norm2_b];
        offset += sizes.v_norm2_b;

        self.v_norm_out_w = data[offset .. offset + sizes.v_norm_out_w];
        offset += sizes.v_norm_out_w;

        self.v_norm_out_b = data[offset .. offset + sizes.v_norm_out_b];
        offset += sizes.v_norm_out_b;

        self.v_proj_fc1_w = data[offset .. offset + sizes.v_proj_fc1_w];
        offset += sizes.v_proj_fc1_w;

        self.v_proj_fc1_b = data[offset .. offset + sizes.v_proj_fc1_b];
        offset += sizes.v_proj_fc1_b;

        self.v_proj_fc2_w = data[offset .. offset + sizes.v_proj_fc2_w];
        offset += sizes.v_proj_fc2_w;

        self.v_proj_fc2_b = data[offset .. offset + sizes.v_proj_fc2_b];
        offset += sizes.v_proj_fc2_b;

        std.debug.print("Mapped weights and biases successfully. \n", .{});
        return self;
    }

    fn calculateSizes(config: Config) struct {
        //text
        word_token_embedding: usize,
        t_ln_w: usize,
        t_ln_b: usize,
        t_Wqkv_w: usize,
        t_Wqkv_b: usize,
        t_out_proj_w: usize,
        t_out_proj_bias: usize,
        t_fc1_w: usize,
        t_fc1_b: usize,
        t_fc2_w: usize,
        t_fc2_b: usize,
        t_linear_w: usize,
        t_linear_b: usize,
        t_ln_out_w: usize,
        t_ln_out_b: usize,
        // vision
        v_patch_embedding_linear_w: usize,
        v_patch_embedding_linear_b: usize,
        v_pos_embedding: usize,
        v_Wqkv_w: usize,
        v_Wqkv_b: usize,
        v_out_proj_w: usize,
        v_out_proj_b: usize,
        v_fc1_w: usize,
        v_fc1_b: usize,
        v_fc2_w: usize,
        v_fc2_b: usize,
        v_norm1_w: usize,
        v_norm1_b: usize,
        v_norm2_w: usize,
        v_norm2_b: usize,
        v_norm_out_w: usize,
        v_norm_out_b: usize,
        v_proj_fc1_w: usize,
        v_proj_fc1_b: usize,
        v_proj_fc2_w: usize,
        v_proj_fc2_b: usize,
    } {
        return .{
            // Text model
            .word_token_embedding = config.dim * config.vocab,
            // Transformer block begins here //
            .t_ln_w = config.n_layers * config.dim,
            .t_ln_b = config.n_layers * config.dim,
            .t_Wqkv_w = config.n_layers * config.dim * config.n_heads * config.head_dim * 3,
            .t_Wqkv_b = config.n_layers * config.n_heads * config.head_dim * 3,
            .t_out_proj_w = config.n_layers * config.dim * config.seq_len,
            .t_out_proj_bias = config.n_layers * config.dim,
            .t_fc1_w = config.n_layers * config.dim * config.hidden_dim,
            .t_fc1_b = config.n_layers * config.hidden_dim,
            .t_fc2_w = config.n_layers * config.hidden_dim * config.dim,
            .t_fc2_b = config.n_layers * config.dim,
            // Transformer block ends here //

            .t_linear_w = config.dim * config.vocab,
            .t_linear_b = config.vocab,
            .t_ln_out_w = config.dim,
            .t_ln_out_b = config.dim,

            //Vision model
            .v_patch_embedding_linear_w = config.patch_size * config.patch_size * config.img_channels * config.vit_dim,
            .v_patch_embedding_linear_b = config.vit_dim,
            .v_pos_embedding = config.vit_dim * ((config.img_dim / config.patch_size) * (config.img_dim / config.patch_size)) * 1,

            // ViT block begins here //
            .v_Wqkv_w = config.n_vit_layers * config.vit_dim * config.n_vit_heads * config.vit_head_dim * 3,
            .v_Wqkv_b = config.n_vit_layers * config.n_vit_heads * config.vit_head_dim * 3,
            .v_out_proj_w = config.n_vit_layers * config.vit_dim * config.vit_dim,
            .v_out_proj_b = config.n_vit_layers * config.vit_dim,
            .v_fc1_w = config.n_vit_layers * config.vit_dim * config.hidden_features,
            .v_fc1_b = config.n_vit_layers * config.hidden_features,
            .v_fc2_w = config.n_vit_layers * config.hidden_features * config.vit_dim,
            .v_fc2_b = config.n_vit_layers * config.vit_dim,
            .v_norm1_w = config.n_vit_layers * config.vit_dim,
            .v_norm1_b = config.n_vit_layers * config.vit_dim,
            .v_norm2_w = config.n_vit_layers * config.vit_dim,
            .v_norm2_b = config.n_vit_layers * config.vit_dim,
            // ViT block ends here //

            .v_norm_out_w = config.vit_dim,
            .v_norm_out_b = config.vit_dim,
            .v_proj_fc1_w = config.vit_head_dim * config.n_heads * config.hidden_dim, //TODO FIGURE OUT WHY THIS IS??
            .v_proj_fc1_b = config.hidden_dim,
            .v_proj_fc2_w = config.hidden_dim * config.dim,
            .v_proj_fc2_b = config.dim,
        };
    }

    fn calculateTotalSize(sizes: anytype) usize {
        var total: usize = 0;
        inline for (std.meta.fields(@TypeOf(sizes))) |field| {
            total += @field(sizes, field.name);
        }
        return total;
    }

    fn deinit(self: *Self, allocator: Allocator) void {
        inline for (std.meta.fields(Self)) |f| {
            allocator.free(@field(self, f.name));
        }
        self.* = undefined;
    }
};

File: weights.py
import numpy as np
from safetensors import safe_open
import struct
import os
import requests
from tqdm import tqdm
from huggingface_hub import hf_hub_download

# Magic number and version constants
WEIGHTS_MAGIC = 0x4D4F4F4E  # "MOON" in ASCII
WEIGHTS_VERSION = 2  # Version 2 for tensor format

def download_file(url, filename):
    """Download file with progress bar"""
    response = requests.get(url, stream=True)
    total_size = int(response.headers.get('content-length', 0))
    
    with open(filename, 'wb') as file, tqdm(
            desc=filename,
            total=total_size,
            unit='B',
            unit_scale=True,
            unit_divisor=1024,
        ) as bar:
        for data in response.iter_content(chunk_size=1024):
            size = file.write(data)
            bar.update(size)
    
    print(f"Downloaded {filename}")

def download_from_hf(repo_id, filename, local_path=None):
    """Download file from Hugging Face"""
    if local_path is None:
        local_path = filename
    
    print(f"Downloading {filename} from {repo_id}...")
    hf_hub_download(repo_id=repo_id, filename=filename, local_dir=".", local_dir_use_symlinks=False)
    
    # Some files might be in subdirectories, so check and move if needed
    if not os.path.exists(local_path) and os.path.exists(os.path.join(repo_id.split('/')[-1], filename)):
        os.rename(os.path.join(repo_id.split('/')[-1], filename), local_path)
        
    print(f"Downloaded {filename}")

def ensure_model_files():
    """Ensure model.safetensors and tokenizer.json files exist"""
    # Define the repository ID for Moondream model
    repo_id = "vikhyatk/moondream2"
    
    # Check and download model.safetensors
    if not os.path.exists("model.safetensors"):
        try:
            download_from_hf(repo_id, "model.safetensors")
        except Exception as e:
            print(f"Error downloading from Hugging Face: {e}")
            # Fallback to direct URL if needed
            model_url = "https://huggingface.co/vikhyatk/moondream2/resolve/main/model.safetensors"
            download_file(model_url, "model.safetensors")
    else:
        print("model.safetensors already exists")
    
    # Check and download tokenizer.json
    if not os.path.exists("tokenizer.json"):
        try:
            download_from_hf(repo_id, "tokenizer.json")
        except Exception as e:
            print(f"Error downloading from Hugging Face: {e}")
            # Fallback to direct URL if needed
            tokenizer_url = "https://huggingface.co/vikhyatk/moondream2/resolve/main/tokenizer.json"
            download_file(tokenizer_url, "tokenizer.json")
    else:
        print("tokenizer.json already exists")

def debug_f16_bits(name: str, value: np.float16):
    bits = value.view(np.uint16)
    sign = (bits & 0x8000) >> 15
    exp = (bits & 0x7C00) >> 10 
    mantissa = bits & 0x03FF
    print(f"{name}: value={value}, bits=0x{bits:04X}")
    print(f"  sign={sign}, exp={exp}, mantissa=0x{mantissa:03X}")

def write_tensor(f, name, tensor, transpose=False):
    """Write a single tensor with its header and data."""
    # Convert to numpy array while verifying it stays as float16
    np_array = tensor.numpy() if hasattr(tensor, 'numpy') else tensor
    assert np_array.dtype == np.float16, f"Expected float16 tensor but got {np_array.dtype}"

    if name == "word_token_embedding":
        print("\nDEBUG: word_token_embedding first 5 values:")
    flat_array = np_array.flatten()
    for i in range(5):
        value = flat_array[i]
        debug_f16_bits(f"Value {i+1}", value)
        print(f"  Raw bytes: {value.tobytes().hex()}")
        print()
        
    # Also print as original tensor to verify no changes
    print("\nOriginal tensor values:")
    print(tensor[:5])
    
    if transpose:
        original_shape = np_array.shape
        np_array = np_array.T
        print(f"Transposing {name} from {original_shape} to {np_array.shape}")
    
    # Get sizes for header
    name_bytes = name.encode('utf-8')
    data_size = np.prod(np_array.shape)
    
    # Write header
    header = struct.pack('<III', 
        len(name_bytes),      
        len(np_array.shape),  
        data_size            
    )
    f.write(header)
    f.write(name_bytes)
    
    # Write shape
    shape_bytes = struct.pack(f'<{len(np_array.shape)}Q', *np_array.shape)
    f.write(shape_bytes)
    
    # Write data
    data_bytes = np_array.tobytes()
    f.write(data_bytes)
    
    current_pos = f.tell()
    print(f"Wrote tensor {name}")
    print(f"  Shape: {np_array.shape}")
    print(f"  dtype: {np_array.dtype}")
    print(f"  Header size: {len(header)}")
    print(f"  Name size: {len(name_bytes)}")
    print(f"  Shape info size: {len(shape_bytes)}")
    print(f"  Data size: {len(data_bytes)}")
    print(f"  Current file position: {current_pos}\n")

def collect_and_stack_layer_tensors(f, n_layers, prefix_template, components):
    """Collect tensors across layers and stack them."""
    # Initialize lists to collect tensors
    collected = {name: [] for name in components.keys()}
    
    # Collect tensors from each layer
    for layer in range(n_layers):
        prefix = prefix_template.format(layer)
        for comp_name, (comp_path, should_transpose) in components.items():
            full_path = f"{prefix}.{comp_path}"
            tensor = f.get_tensor(full_path).numpy()
            # Transpose weight matrices before stacking if needed
            if should_transpose:
                tensor = tensor.T
            collected[comp_name].append(tensor)
    
    # Stack tensors along a new first dimension
    stacked = {name: np.stack(tensors) for name, tensors in collected.items()}
    return stacked

def convert_safetensors_to_binary(input_file, output_file):
    with safe_open(input_file, framework="pt", device="cpu") as f:
        with open(output_file, 'wb') as out_f:
            # Write magic number and version
            out_f.write(struct.pack('<II', WEIGHTS_MAGIC, WEIGHTS_VERSION))
            
            # Write word embedding (no transpose)
            write_tensor(out_f, "word_token_embedding", 
                        f.get_tensor("model.text.wte"))

            # Collect and write text transformer layer tensors
            text_components = {
                # (path, should_transpose)
                "t_ln_w": ("ln.weight", False),
                "t_ln_b": ("ln.bias", False),
                "t_Wqkv_w": ("attn.qkv.weight", False),
                "t_Wqkv_b": ("attn.qkv.bias", False),
                "t_out_proj_w": ("attn.proj.weight", False),
                "t_out_proj_bias": ("attn.proj.bias", False),
                "t_fc1_w": ("mlp.fc1.weight", False),  # Changed to False
                "t_fc1_b": ("mlp.fc1.bias", False),
                "t_fc2_w": ("mlp.fc2.weight", False),  # Changed to False
                "t_fc2_b": ("mlp.fc2.bias", False),
            }
            
            text_stacked = collect_and_stack_layer_tensors(
                f, 24, "model.text.blocks.{}", text_components
            )
            
            # Write stacked text tensors
            for name, tensor in text_stacked.items():
                write_tensor(out_f, name, tensor)

            # Write text model end layers (with transpose for weights)
            write_tensor(out_f, "t_linear_w", 
                        f.get_tensor("model.text.lm_head.weight"), transpose=False)
            write_tensor(out_f, "t_linear_b", 
                        f.get_tensor("model.text.lm_head.bias"))
            write_tensor(out_f, "t_ln_out_w", 
                        f.get_tensor("model.text.post_ln.weight"))
            write_tensor(out_f, "t_ln_out_b", 
                        f.get_tensor("model.text.post_ln.bias"))

            # Vision model start layers
            write_tensor(out_f, "v_patch_embedding_linear_w",
                        f.get_tensor("model.vision.patch_emb.weight"), transpose=True)
            write_tensor(out_f, "v_patch_embedding_linear_b",
                        f.get_tensor("model.vision.patch_emb.bias"))
            write_tensor(out_f, "v_pos_embedding",
                        f.get_tensor("model.vision.pos_emb"))

            # Collect and write vision transformer layer tensors
            vision_components = {
                # (path, should_transpose)
                "v_Wqkv_w": ("attn.qkv.weight", True),
                "v_Wqkv_b": ("attn.qkv.bias", False),
                "v_out_proj_w": ("attn.proj.weight", True),
                "v_out_proj_b": ("attn.proj.bias", False),
                "v_fc1_w": ("mlp.fc1.weight", True), 
                "v_fc1_b": ("mlp.fc1.bias", False),
                "v_fc2_w": ("mlp.fc2.weight", True),  
                "v_fc2_b": ("mlp.fc2.bias", False),
                "v_norm1_w": ("ln1.weight", False),
                "v_norm1_b": ("ln1.bias", False),
                "v_norm2_w": ("ln2.weight", False),
                "v_norm2_b": ("ln2.bias", False),
            }
            
            vision_stacked = collect_and_stack_layer_tensors(
                f, 27, "model.vision.blocks.{}", vision_components
            )
            
            # Write stacked vision tensors
            for name, tensor in vision_stacked.items():
                write_tensor(out_f, name, tensor)

            # Vision end layers
            write_tensor(out_f, "v_norm_out_w",
                        f.get_tensor("model.vision.post_ln.weight"))
            write_tensor(out_f, "v_norm_out_b",
                        f.get_tensor("model.vision.post_ln.bias"))

            # Projection layers
            write_tensor(out_f, "v_proj_fc1_w",
                        f.get_tensor("model.vision.proj_mlp.fc1.weight"), transpose=True)
            write_tensor(out_f, "v_proj_fc1_b",
                        f.get_tensor("model.vision.proj_mlp.fc1.bias"))
            write_tensor(out_f, "v_proj_fc2_w",
                        f.get_tensor("model.vision.proj_mlp.fc2.weight"), transpose=True)
            write_tensor(out_f, "v_proj_fc2_b",
                        f.get_tensor("model.vision.proj_mlp.fc2.bias"))

def delete_file(file_path):
    """Delete a file if it exists and return success status"""
    if os.path.exists(file_path):
        try:
            os.remove(file_path)
            return True
        except Exception as e:
            print(f"Error deleting {file_path}: {e}")
            return False
    else:
        print(f"File {file_path} does not exist")
        return False

if __name__ == "__main__":
    # First ensure we have the model files
    print("Checking for required model files...")
    ensure_model_files()
    
    # Now run the conversion
    input_file = 'model.safetensors'
    output_file = 'moondream.bin'
    print(f"\nConverting {input_file} to {output_file}...")
    convert_safetensors_to_binary(input_file, output_file)
    
    print(f"\nConversion complete! Output saved to {output_file}")
    
    # Delete the original safetensors file to save space
    print("\nDeleting original safetensors file to save disk space...")
    if delete_file(input_file):
        print(f"Successfully deleted {input_file}")
    
    print("\nProcess complete! The converted model is available as moondream.bin and tokenizer.json is preserved.")